apps/rule-manager/client/src/ts/components/Diff.tsx

```diff
@@ -0,0 +1,163 @@
+    const draftFieldObjects = Object.keys(rule.draft).map(fieldName => { return { fieldName, value: rule.draft[fieldName as keyof BaseRule]}});
+    const liveFieldObjects = Object.keys(newestLiveRule).map(fieldName => { return { fieldName, value: newestLiveRule[fieldName as keyof BaseRule]}})
+
+    const fieldsToProvideDiffFor = [...textDiffFields, ...comparisonDiffFields];
+    // The rule data contains various fields we're not interested in providing diffs for
+    const draftFieldsToDiff = draftFieldObjects.filter(draftField => fieldsToProvideDiffFor.includes(draftField.fieldName));
+    const liveFieldsToDiff = liveFieldObjects.filter(liveField => fieldsToProvideDiffFor.includes(liveField.fieldName));
+    
+    // Add fields that only appear in one of live and draft
+    const nonIntersectingFields = findNonIntersectingFields(draftFieldsToDiff, liveFieldsToDiff);
+
+    // Compare fields that appear in both live and draft
+    const divergentFields = draftFieldsToDiff.reduce((divergentFieldsArray, draftField) => {
+        const equivalentLiveField = liveFieldsToDiff.find(liveField => liveField.fieldName === draftField.fieldName)
+        if (equivalentLiveField){
+            const liveAndDraftValuesMatch = doValuesMatch(draftField.value, equivalentLiveField.value);
+            if (!liveAndDraftValuesMatch){
+                divergentFieldsArray.push({
+                    fieldName: draftField.fieldName,
+                    live: equivalentLiveField.value,
+                    draft: draftField.value
+                })
+            }
+        }
+        return divergentFieldsArray
+    }, nonIntersectingFields)
+    return divergentFields;
+};
+
+export const transformToHumanReadableValues = (fields: DivergentField[], tags: Record<number, Tag>) => {
+    return fields.map(field => {
+        // Use tag names instead of ids
+        if (field.fieldName === "tags"){
+            if (Array.isArray(field.draft)){
+                field.draft = field.draft.map(field => tags[field as number] ? tags[field as number].name : field.toString())
+            }
+            if (Array.isArray(field.live)){
+                field.live = field.live.map(field => tags[field as number] ? tags[field as number].name : field.toString() )
+            }
+        }
+        // Use readable ruleType names
+        if (field.fieldName === "ruleType"){
+            field.draft = ruleTypeOptions.find(option => option.id === field.draft)?.label || field.draft
+            field.live = ruleTypeOptions.find(option => option.id === field.live)?.label || field.live
+        }
+        return field
+    })
+}
+
+export const Diff = ({rule}: {rule: RuleData | undefined}) => {
+    const {tags} = useTags();
+    
+    const diffedFields = rule ? transformToHumanReadableValues(findFieldsWithDiffs(rule), tags) : [];
+    const textDiffs = diffedFields.filter(diffedField => textDiffFields.includes(diffedField.fieldName))
```

Comment:
Non-blocking – I'd be tempted to move the logic that cared about what sort of diff to provide into a `<FieldDiff` object, which then decided which sort of diff to render. It'd make this component simpler – you could even pass the `DivergentField` straight to the component – which'd then decide which sort of diff to render, and keep all of the diff-related logic together.